src/Avalonia.Base/Data/Core/MultiBindingExpression.cs

```diff
@@ -0,0 +1,131 @@
+            : base(priority)
+    {
+        _bindings = [.. bindings];
+        _converter = converter;
+        _converterCulture = converterCulture;
+        _converterParameter = converterParameter;
+        _expressions = new UntypedBindingExpressionBase[_bindings.Length];
+        _fallbackValue = fallbackValue;
+        _targetNullValue = targetNullValue;
+        _values = new object?[_bindings.Length];
+        _valuesView = new(_values);
+
+#if NETSTANDARD2_0
+        for (var i = 0; i < _bindings.Length; ++i)
+            _values[i] = s_uninitialized;
+#else
+        Array.Fill(_values, s_uninitialized);
+#endif
+    }
+
+    public override string Description => "MultiBinding";
+
+    protected override void StartCore()
+    {
+        if (!TryGetTarget(out var target))
+            throw new AvaloniaInternalException("MultiBindingExpression has no target.");
+
+        for (var i = 0; i < _bindings.Length; ++i)
+        {
+            var binding = _bindings[i]; 
+
+            if (binding is not IBinding2 b)
+                throw new NotSupportedException($"Unsupported IBinding implementation '{binding}'.");
+
+            var expression = b.Instance(target, null, null);
+
+            if (expression is not UntypedBindingExpressionBase e)
+                throw new NotSupportedException($"Unsupported BindingExpressionBase implementation '{expression}'.");
+
+            _expressions[i] = e;
+            e.AttachAndStart(this, target, null, Priority);
+        }
+    }
+
+    protected override void StopCore()
+    {
+        for (var i = 0; i < _expressions.Length; ++i)
+        {
+            _expressions[i]?.Dispose();
+            _expressions[i] = null;
+            _values[i] = s_uninitialized;
+        }
+    }
+
+    void IBindingExpressionSink.OnChanged(
+        UntypedBindingExpressionBase instance,
+        bool hasValueChanged,
+        bool hasErrorChanged,
+        object? value,
+        BindingError? error)
+    {
+        var i = Array.IndexOf(_expressions, instance);
+        Debug.Assert(i != -1);
+
+        _values[i] = BindingNotification.ExtractValue(value);
+        PublishValue();
+    }
+
+    void IBindingExpressionSink.OnCompleted(UntypedBindingExpressionBase instance)
+    {
+        // Nothing to do here.
+    }
+
+    private void PublishValue()
+    {
+        foreach (var v in _values)
+        {
+            if (v == s_uninitialized)
+                return;
+        }
+
+        var culture = _converterCulture ?? CultureInfo.CurrentCulture;
+
+        if (_converter is not null)
+        {
+            var converted = _converter.Convert(_valuesView, TargetType, _converterParameter, culture);
+
+            converted = BindingNotification.ExtractValue(converted);
+
+            if (converted == null)
+                converted = _targetNullValue;
+            if (converted == AvaloniaProperty.UnsetValue)
+                converted = _fallbackValue;
+            PublishValue(converted);
```

Comment:
This doesn't seem to handle `BindingOperations.DoNothing` anymore: `DoNothing` is directly published to the target.

`MultiValueConverterTests` is supposed to check this but the test passes for the wrong reason, because `DoNothingType` isn't convertible to the target `string` type.

Reply:
Good catch! Fixed the test and the issue.